core/internal/api/credentials.go

```diff
@@ -55,3 +72,240 @@ func (c *apiKeyCredentialProvider) Apply(req *http.Request) error {
+type ExpiresAt time.Time
+
+const expiresAtLayout = "2006-01-02 15:04:05"
+
+func (e *ExpiresAt) UnmarshalJSON(data []byte) error {
+	var timeString string
+	if err := json.Unmarshal(data, &timeString); err != nil {
+		return err
+	}
+
+	parsedTime, err := time.Parse(expiresAtLayout, timeString)
+	if err != nil {
+		return err
+	}
+
+	*e = ExpiresAt(parsedTime)
+	return nil
+}
+
+func (e ExpiresAt) MarshalJSON() ([]byte, error) {
+	formattedTime := time.Time(e).Format(expiresAtLayout)
+	return json.Marshal(formattedTime)
+}
+
+type accessTokenInfo struct {
+	// The time at which the access token will expire
+	ExpiresAt ExpiresAt `json:"expires_at"`
+
+	// The access token to use for authentication
+	AccessToken string `json:"access_token"`
+}
+
+func (c *accessTokenInfo) IsTokenExpiring() bool {
+	return time.Until(time.Time(c.ExpiresAt)) <= time.Minute*5
+}
+
+// CredentialsFile is used when serializing/deserializing JSON data from the
+// credentials file
+type CredentialsFile struct {
+	Credentials map[string]accessTokenInfo `json:"credentials"`
+}
+
+// Apply checks if the access token is expiring, and fetches a new one if so.
+// It then supplies the access token to the request via the Authorization header
+// as a Bearer token
+func (c *oauth2CredentialProvider) Apply(req *http.Request) error {
+	if c.shouldRefreshToken() {
+		err := c.loadCredentials()
+		if err != nil {
+			return err
+		}
+	}
+
+	req.Header.Set(
+		"Authorization",
+		"Bearer "+c.tokenInfo.AccessToken,
+	)
+	return nil
+}
+
+func (c *oauth2CredentialProvider) shouldRefreshToken() bool {
+	c.tokenMu.RLock()
+	defer c.tokenMu.RUnlock()
+
+	return c.tokenInfo.IsTokenExpiring()
+}
+
+// loadCredentials ensures the access token is valid by refreshing it if
+// necessary, using a mutex to prevent concurrent refreshes. It first checks for
+// a non-expiring token in memory or the credentials file. If none is found, it
+// fetches a new token and saves it.
+func (c *oauth2CredentialProvider) loadCredentials() error {
+	c.tokenMu.Lock()
+	defer c.tokenMu.Unlock()
+
+	// if the access token has already been refreshed, return early
+	if !c.tokenInfo.IsTokenExpiring() {
+		return nil
+	}
+
+	credsFile, ok := c.tryLoadCredentialsFromFile()
+	if ok {
+		accessToken, ok := credsFile.Credentials[c.baseURL]
+		if ok && !accessToken.IsTokenExpiring() {
+			c.tokenInfo = accessToken
+			return nil
+		}
+	}
+
+	token, err := c.fetchAccessToken()
+	if err != nil {
+		return fmt.Errorf("api: couldn't fetch access token: %v", err)
+	}
+	c.tokenInfo = token
+
+	c.trySaveCredentialsToFile(credsFile)
+
+	return nil
+}
+
+// tryLoadCredentialsFromFile attempts to load the access token from the credentials file
+func (c *oauth2CredentialProvider) tryLoadCredentialsFromFile() (CredentialsFile, bool) {
+	var credsFile CredentialsFile
+
+	file, err := os.ReadFile(c.credentialsFilePath)
+	if err != nil {
+		c.logger.Warn("failed to read credentials file",
```

Comment:
all logged messages in this file should be prefixed with `api: `, so

```suggestion
		c.logger.Warn("api: failed to read credentials file",
```